torch/utils/data/datapipes/iter/combining.py

```diff
@@ -137,24 +135,70 @@ def __iter__(self):
 
 @functional_datapipe('demux')
 class DemultiplexerIterDataPipe(IterDataPipe):
+    r""" :class:`DemultiplexerIterDataPipe`.
+
+        Iterable DataPipe to split the input DataPipe into multiple child DataPipes, using the given
+        classification function. A list of the child DataPipes is returned from this operation.
+        args:
+            datapipe: Iterable DataPipe being filtered
+            num_instances: number of instances of the DataPipe to create
+            classifier_fn: a function that maps values to an integer within the range [0, num_instances - 1]
+            buffer_size: this defines the maximum number of inputs that the buffer can hold across all child
+             DataPipes while waiting for their values to be yielded
+    """
+    def __new__(cls, datapipe: IterDataPipe, num_instances: int,
+                classifier_fn: Callable[[T_co], int], buffer_size: int = 1000):
+        container = _DemultiplexerIterDataPipe(datapipe, num_instances, classifier_fn, buffer_size)
+        return [ChildDataPipe(container, i) for i in range(num_instances)]
+
+    def __init__(self, *arg):
+        raise Exception("__init__ called instead of __new__")
+
 
-    # given n = num_instances
-    # You want classifier_fn to return (0, ..., n - 1)
-    # Check and make sure the output from classifier_fn is within range [0, n-1]
-    # We are forcing people to return an int
+class _DemultiplexerIterDataPipe(IterDataPipe):
+    r""" :class:`_DemultiplexerIterDataPipe`.
 
-    def __init__(self, datapipe: IterDataPipe[T_co], classifier_fn: Callable[[T_co], int]):
-        self.datapipe = datapipe
+        Container to hold instance-specific information on behalf of DemultiplexerIterDataPipe. It tracks
+        the state of its child DataPipes, maintains the buffer, classifies and yields the next correct value
+        as requested by the child DataPipes.
+    """
+
+    def __init__(self, datapipe: IterDataPipe[T_co], num_instances: int,
+                 classifier_fn: Callable[[T_co], int], buffer_size: int):
+        self.main_datapipe = iter(datapipe)
```

Comment:
drop `iter` please

Reply:
As discussed in the other PR, I need `iter` for the subsequent `__next__` call.